### DISCUSSION TITLE
Perl files misidentified as "Raku"

### DISCUSSION THREAD

**Discussion by mauke** [github-linguist/linguist#6263] [Classification] [answered]
https://github.com/mauke/HTML-Blitz/search?l=raku shows that three of my Perl modules are misidentified as "Raku":

- https://github.com/mauke/HTML-Blitz/blob/34301669b77ad6640dcabb65ab3b1ff7a860e585/lib/HTML/Blitz/ParseError.pm
- https://github.com/mauke/HTML-Blitz/blob/34301669b77ad6640dcabb65ab3b1ff7a860e585/lib/HTML/Blitz/Parser.pm
- https://github.com/mauke/HTML-Blitz/blob/34301669b77ad6640dcabb65ab3b1ff7a860e585/lib/HTML/Blitz/Atom.pm

I don't understand why. From https://github.com/github/linguist/blob/c1c34e5260797b4d598f5ec76f19723bfc5a1894/lib/linguist/heuristics.yml#L815 it seems like it's looking for lines starting with `use v6`, `module`, `class`, or `my class`. But none of these apply to my files.

**Comment by lildude** [comment, marked as answer]
This will be because your files don't match [any of the heuristics for the `.pm` extension](https://github.com/github/linguist/blob/c1c34e5260797b4d598f5ec76f19723bfc5a1894/lib/linguist/heuristics.yml#L509-L516) so it falls through to the "last resort" classifier which will make a guess based on the samples we have for all languages that use this extension. This is far from perfect and in this case it's got it wrong three times; ie it thinks your files look more like the Raku samples than the Perl samples.

Improving the Perl heuristic would be ideal, but I doubt there's more specificity that can be added which wouldn't also apply to Raku (from my limited understanding of both languages). Adding more samples might help, but it might not given how similar Perl is to Raku. 

There are two solutions you can implement:

1. add `use strict;` or `use <5 VERSION>;` or similar to the files to force the use of Perl 5. This will be picked up by the Perl 5 heuristic
2. force the language with the `linguist-language=Perl` language override.

**Reply by mauke** [reply]
I think the following patterns could be added for Perl:

- `^\s*use\s+constant\b` because Raku does constants differently
- `^\s*use\s+overload\b` because Raku does overloading differently
- `^\s*(?:our\s*)?@EXPORT\s*=` because Raku does exporting differently
- `^\s*package\s+\w+(?:::\w+)*\s*[;{]` because Raku does packages differently
- `\w->[a-zA-Z_\[({]` because Raku does method calls and references differently

**Reply by lildude** [reply]
You certainly know a lot more than me so please feel free to submit a PR with these changes 😁. 

I think we might also be able to simplify some of those a bit too, but we'll leave that for the PR.

**Reply by mauke** [reply]
Cool. I created PR #6264. Hope it makes sense?

**Comment by bessarabov** [comment]
Maybe it is worth always consider that the files with .pm extension is perl5? This will remove all the false positive when perl5 is incorrectly detected as raku.

Such change will make raku detection worse, but the official extension for raku is .rakumod, and probably if the raku code is using .pm extension it is best to rename the file to have extension .rakumod
